File: android/digia-engage/src/main/java/com/digia/engage/internal/model/NudgeConfig.kt
package com.digia.engage.internal.model

import org.json.JSONObject

internal enum class NudgeDisplayType {
    BOTTOM_SHEET,
    DIALOG;

    companion object {
        fun fromString(value: String?): NudgeDisplayType =
            if (value == "dialog") DIALOG else BOTTOM_SHEET
    }

    val displayStyle: String
        get() = when (this) {
            BOTTOM_SHEET -> "bottom_sheet"
            DIALOG -> "dialog"
        }
}

internal data class NudgeSurface(
    val displayType: NudgeDisplayType = NudgeDisplayType.BOTTOM_SHEET,
    val backgroundColor: String? = null,
    val barrierColor: String? = null,
    val cornerRadius: Float = 18f,
    val padding: Float = 20f,
    val backdropDismissible: Boolean = true,
    val showCloseButton: Boolean = false,
    val showHandle: Boolean = true,
    val draggable: Boolean = true,
    val widthFraction: Float = 0.86f,
) {
    companion object {
        fun fromJson(json: JSONObject?): NudgeSurface {
            val j = json ?: JSONObject()
            val widthPct = j.optDouble("widthPct", 86.0)
            return NudgeSurface(
                displayType = NudgeDisplayType.fromString(j.optString("displayType")),
                backgroundColor = j.optString("backgroundColor").ifBlank { null },
                barrierColor = j.optString("barrierColor").ifBlank { null },
                cornerRadius = j.optDouble("cornerRadius", 18.0).toFloat(),
                padding = j.optDouble("padding", 20.0).toFloat(),
                backdropDismissible = j.optBoolean("backdropDismissible", true),
                showCloseButton = j.optBoolean("showCloseButton", false),
                showHandle = j.optBoolean("showHandle", true),
                draggable = j.optBoolean("draggable", true),
                widthFraction = (widthPct / 100.0).toFloat().coerceIn(0.3f, 1.0f),
            )
        }
    }
}

internal data class NudgeConfig(
    val surface: NudgeSurface,
    val content: NudgeColumn,
) {
    companion object {
        fun fromJson(json: JSONObject): NudgeConfig? = NudgeParser().parse(json)
    }
}

File: flutter/lib/api/internal/nudge/nudge_config.dart
import 'dart:ui' show Color;

import 'nudge_content.dart';

/// How a nudge surface presents over the host app.
enum NudgeDisplayType { bottomSheet, dialog }

/// The presentation chrome for a nudge — everything *around* the content tree.
///
/// A pure value object; decoding from the `container` wire object lives in
/// `nudge_parser.dart`. The content layout (spacing / alignment) lives on the
/// [NudgeColumn], so this only describes how the modal frame looks and behaves.
class NudgeSurface {
  final NudgeDisplayType displayType;

  /// Surface background; null inherits white at render time.
  final Color? backgroundColor;
  final double cornerRadius;

  /// Uniform inner padding around the content tree, in logical pixels.
  final double padding;

  /// Dismiss when the scrim/barrier outside the surface is tapped.
  final bool backdropDismissible;

  /// Render an "×" close affordance on the surface.
  final bool showCloseButton;

  /// Show the drag-handle pill at the top of the sheet (bottom sheet only).
  final bool showHandle;

  /// Allow dragging the sheet down to dismiss (bottom sheet only).
  final bool draggable;

  /// Dialog width as a fraction of the screen width, 0…1 (dialog only).
  final double widthFraction;

  const NudgeSurface({
    required this.displayType,
    required this.backgroundColor,
    required this.cornerRadius,
    required this.padding,
    required this.backdropDismissible,
    required this.showCloseButton,
    required this.showHandle,
    required this.draggable,
    required this.widthFraction,
  });

  bool get isBottomSheet => displayType == NudgeDisplayType.bottomSheet;
}

/// A fully parsed nudge: the presentation [surface] plus the typed content
/// tree ([content]) the renderer draws inside it.
class NudgeConfig {
  final NudgeSurface surface;
  final NudgeColumn content;

  const NudgeConfig({required this.surface, required this.content});
}
